### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by ArtemSydun**

### CODE REVIEW

File: src/controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  res.status(statusCode.CREATED).json({ status: "success" });
+};
+
+const loginUserController = async (req, res) => {
+  const body = req.body;
+  const token = await loginUser(body);
+  res.status(statusCode.OK).json({ status: "success", token });
+};
+
+const getUserInfoController = async (req, res) => {
```
Comment (yurii-shendryk) [inline src/controllers/authController.js:23]: This is not related to auth, better to create another controller

File: src/controllers/authController.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  const { email } = req.body;
+  const userInfo = await getUserInfo(email);
+  res.status(statusCode.OK).json(userInfo);
+};
+
+const addToFavoritesController = async (req, res) => {
```
Comment (yurii-shendryk) [inline src/controllers/authController.js:29]: This is not related to auth, better to create another controller

File: src/controllers/authController.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  const { email, movieId } = req.body;
+
+  const updatedFavorites = await addToFavorites(email, movieId);
+  res.status(statusCode.OK).send(updatedFavorites);
+};
+
+const getUserFavoriteMoviesController = async (req, res) => {
```
Comment (yurii-shendryk) [inline src/controllers/authController.js:36]: This is not related to auth, better to create another controller

File: src/controllers/authController.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  const { email } = req.body;
+
+  const favoriteMovies = await getAllFavorites(email);
+
+  res.status(statusCode.OK).json(favoriteMovies);
+}
+
+const clearUserFavoritesController = async (req, res) => {
```
Comment (yurii-shendryk) [inline src/controllers/authController.js:44]: This is not related to auth, better to create another controller

File: src/db/connection.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+const mongoose = require('mongoose');
+
+const MONGO_DB_CONNECTION_STRING = process.env.MONGO_ATLAS_CONNECTION_STRING;
+const MONGO_DB_NAME='Movies-libraries';
```
Comment (yurii-shendryk) [inline src/db/connection.js:4]: It would be better to store this value in the .env

File: src/helpers/apiWrapper.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+const asyncWrapper = controller => {
+  return async (req, res, next) => {
+    try {
+      await controller(req, res);
+    } catch (error) {
+      console.log({ ERRROR: error.message });
```
Comment (yurii-shendryk) [inline src/helpers/apiWrapper.js:6]: you can remove this console.log()

File: src/controllers/authController.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+const {
+  registerUser,
+  loginUser,
+  getUserInfo,
+  addToFavorites,
+  getAllFavorites,
+  clearFavorites,
+} = require("../services/authService.js");
+
+const registerUserController = async (req, res, next) => {
```
Comment (yurii-shendryk) [inline src/controllers/authController.js:11]: you can omit next fucntion here

File: src/controllers/moviesController.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
+const { statusCode } = require("../helpers/constants.js");
+const {
+  getMovies,
+  getMovieById,
+} = require("../services/moviesService.js");
+
+const getAllMovies = async (req, res, next) => {
```
Comment (yurii-shendryk) [inline src/controllers/moviesController.js:7]: You can omit using next fucntion, but wou have to use asyncWrapper in the appropriate router

File: src/controllers/moviesController.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    const movies = await getMovies();
+
+    res.status(statusCode.OK).send(movies);
+  } catch (err) {
+    next(err)
+  }
+};
+
+
+const getMovie = async (req, res, next) => {
```
Comment (yurii-shendryk) [inline src/controllers/moviesController.js:18]: You can omit using next fucntion, but wou have to use asyncWrapper in the appropriate router

File: src/controllers/moviesController.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  try {
+    const { id } = req.params;
+    const movieById = await getMovieById(id);
+
+    if (!movieById) {
```
Comment (yurii-shendryk) [inline src/controllers/moviesController.js:23]: check comment abowe, you can just throw an error

File: src/routers/authRouter.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+const {
+  validateCreateUser,
+  validateCredentialsUser
+} = require('../middlewares/authValidation.js');
+
+const { authGuard } = require('../middlewares/authGuard.js');
+
+router.get('/current', authGuard, asyncWrapper(getUserInfoController))
+router.get('/favorites', authGuard, asyncWrapper(getUserFavoriteMoviesController))
+router.post('/addToFavorites', authGuard, asyncWrapper(addToFavoritesController))
```
Comment (yurii-shendryk) [inline src/routers/authRouter.js:24]: according to REST you can rename route path to /favorites

File: src/routers/authRouter.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+router.post('/clearFavorites', authGuard, asyncWrapper(clearUserFavoritesController))
```
Comment (yurii-shendryk) [inline src/routers/authRouter.js:25]: You can use patch method and change path to /favorites

File: src/routers/authRouter.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+const {
+  validateCreateUser,
+  validateCredentialsUser
+} = require('../middlewares/authValidation.js');
+
+const { authGuard } = require('../middlewares/authGuard.js');
+
+router.get('/current', authGuard, asyncWrapper(getUserInfoController))
+router.get('/favorites', authGuard, asyncWrapper(getUserFavoriteMoviesController))
```
Comment (yurii-shendryk) [inline src/routers/authRouter.js:23]: this is not related to auth, it is better to create separate router

File: src/routers/authRouter.js (hunk 11 above)
Comment (yurii-shendryk) [inline src/routers/authRouter.js:24]: this is not related to auth, it is better to create separate router

File: src/routers/authRouter.js (hunk 12 above)
Comment (yurii-shendryk) [inline src/routers/authRouter.js:25]: this is not related to auth, it is better to create separate router

File: src/services/authService.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+const addToFavorites = async (email, movieId) => {
+  const user = await User.findOne({ email });
+
+  const isMovieAlreadyAdded = user.favoriteMovies.some(favMovieId => favMovieId.equals(movieId));
+  
+  if (isMovieAlreadyAdded) {
+    throw new ConflictException('Film already in favorites');
+  }
+  
+  await User.updateOne(
```
Comment (yurii-shendryk) [inline src/services/authService.js:63]: you can update finded user from 55 code line

File: src/services/authService.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    { _id: user.id },
+    { $push: { favoriteMovies: movieId } }
+  );
+
+  const updatedUser = await User.findById(user._id).populate('favoriteMovies');
```
Comment (yurii-shendryk) [inline src/services/authService.js:68]: you can use user from 63 code line

File: src/services/authService.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+  return updatedUser.favoriteMovies;
+};
+
+const getAllFavorites = async (email) => {
+  const user = await User.findOne({ email });
+
+  return user.favoriteMovies;
```
Comment (yurii-shendryk) [inline src/services/authService.js:75]: You can populate movies for more detailed info

**Review by yurii-shendryk — APPROVED** [review]
Good job

File: src/controllers/authController.js (hunk 1 above)
Comment (ArtemSydun) [reply src/controllers/authController.js:23]: fixed
